cat-shark-carousel/src/pages/Home.tsx

```diff
@@ -0,0 +1,68 @@
+import React, { useState } from 'react';
+import Button from '../components/button';
+import Carousel, { IChange } from '../components/carousel';
+import Loading from '../components/loading';
+import useGet from '../hooks/useGet';
+import PhotoType from '../types/photoType';
+import { Endpoint } from '../util/url';
+
+const Index = () => {
+    const [state, updateState] = useState({ activeIndex: 0, photoTypes: [PhotoType.CAT, PhotoType.SHARK] })
+    const [getState, updateParams] = useGet({ endpoint: Endpoint.PHOTOS, params: { types: state.photoTypes } })
+
+    const toggle = (photoType: PhotoType) => {
+        let photoTypes: PhotoType[] = [];
+        if (state.photoTypes.includes(photoType)) {
+            photoTypes = state.photoTypes.filter(value => value !== photoType)
+        } else {
+            photoTypes = [...state.photoTypes, ...[photoType]];
+        }
+        updateParams({ params: { types: photoTypes } });
+        updateState({ photoTypes, activeIndex: 0 })
+    };
+
+    const updateIndex = (change: IChange) => {
+        let activeIndex = state.activeIndex + change;
+        if (activeIndex < 0) {
+            activeIndex = getState.data.length - 1;
+        } else if (activeIndex >= getState.data.length) {
+            activeIndex = 0;
+        }
+        updateState(prev => ({ ...prev, ...{ activeIndex } }));
+    };
+
+    return (
+        <main className="flex flex-columns h-100 align-center">
+            <header className="flex align-center flex-columns">
+                <h1>🐈 Pictures of Ferocious Beasts 🦈</h1>
+                <h2>
+                    The ultimate repository, currently showing
+                    {state.photoTypes.length > 0 ? ` ${state.photoTypes[0].toLocaleLowerCase()}s ` : " nothing"}
+                    {state.photoTypes.length > 1 ? ` and ${state.photoTypes[1].toLocaleLowerCase()}s` : ""}
```

Comment:
What would you do here if you needed to support more photo types?

Reply:
This was a fun question that I spend a bit of time on and my solution ended up being a bit larger just so I could test it out. I’ll try and divide this answer into two parts: The answer, and my test to prove / show it.

## Answer

To make the subTitle more generic to support more types of pictures being added in the future, I’d do the following:

Add a `useState` statement at the top of the file to track the text that I am planning on showing to the user.

```typescript
const [text, setText] = useState('');
```

Then, using an `useEffect`, I would watch for changes to my state which tracks what photos I’m currently selected. Inside of the `useEffect` I would create an algorithm that would generate the subText that I would show the user. I coded up a solution as follows:

```typescript
useEffect(() => {
    const newText = state.photoTypes
        .map((photo) => `${photo.toLocaleLowerCase()}s`)
        .reduce((photos: string[], photo, index) => {
            let addition = ',';
            if (state.photoTypes.length === 1) {
                return [photo];
            } else if (state.photoTypes.length === 2) {
                addition = '';
            }
            const text = (index === state.photoTypes.length - 1) ? `and ${photo}` : `${photo}${addition}`;
            return [...photos, ...[text]]
        }, [])
        .join(' ');
    if (newText.length > 0) {
        setText(`The ultimate repository, currently showing ${newText}`);
    } else {
        setText('The ultimate repository, currently showing nothing')
    }
}, [state.photoTypes, setText])
```

The magic happens in the reduce function where I add on pieces of text to the photo text depending on the length of the array. All text gets `,` added to it unless the length of the array is equal to 2 or 1. After I’ve created all of the strings, I join them together with spaces. At the end I set the text to what it should be.

Finally the header is displayed as such:

```typescript
<header className="flex align-center flex-columns">
    <h1>🐈 Pictures of Ferocious Beasts 🦈</h1>
    <h2>{text}</h2>
</header>
```

## Testing

To test out if my solution would work I decided to have some fun and make it easy for me to add more buttons by just adding 2 pieces of code: The photo type and the photo button text. First, I extended my enum to include “Dogs” and “Bees” and created another enum that would contain the text for the buttons:


```typescript
enum PhotoType {
    SHARK = "SHARK",
    CAT = "CAT",
    DOG = "DOG",
    BEE = "BEE"
}

export enum PhotoTypeButtonText {
    SHARK = "🦈 Sharks",
    CAT = "🐈 Cats",
    DOG = "🐕 Dogs",
    BEE = "🐝 Bees"
}
```

After which, I had to update my rendering logic in the home component. Because I wanted to be type safe, I had to do some fun mapping logic over the `PhotoType` enum by mapping over the keys of `PhotoType` and mapping it back into `PhotoType`. This can be seen in the code below. Finally, using the `PhotoType` key, I was easily able to also add in the button text for that particular `PhotoType`:

```typescript
<section className="flex mb-2">
    {Object.keys(PhotoType)
        .map((photoType) => PhotoType[photoType as PhotoType])
        .map((photoType) =>
            <Button active={state.photoTypes.includes(photoType)}
                onClick={() => toggle(photoType)}
            >
                {PhotoTypeButtonText[photoType]}
            </Button>
        )}
</section>
```

Overall, the solution works very well:

<img width="643" alt="The ultimate repository, currently showing bees, dogs," src="https://user-images.githubusercontent.com/12245650/130308721-13d9868f-59d7-4373-ac14-d71f1e445211.png">

<img width="641" alt="The ultimate repository, currently showing cats and" src="https://user-images.githubusercontent.com/12245650/130308724-b6daf226-a306-4a56-a296-880b523e8469.png">

<img width="639" alt="The ultimate repository, currently showing sharks" src="https://user-images.githubusercontent.com/12245650/130308729-87a13ee6-90e0-4bfd-aa48-c680bde01071.png">

<img width="640" alt="The ultimate repository, currently showing nothing" src="https://user-images.githubusercontent.com/12245650/130308732-ef7acadd-0d90-4c77-b93d-2d50dcac6df2.png">

Reply:
Ended up looking at my code this morning and I realized that I didn't need the reduce, so I removed it and replaced it with a `map` call instead. I think when I initially wrote the first piece of code, I was trying to **not** use `join` but ended up adding it in. Because of that addition, there is no reason to use reduce as I'm still returning the array with slight modifications to it.

```typescript
useEffect(() => {
    const newText = state.photoTypes
        .map((photo) => `${photo.toLocaleLowerCase()}s`)
        .map((photo, index) => {
            let addition = ',';
            if (state.photoTypes.length === 1) {
                return photo;
            } else if (state.photoTypes.length === 2) {
                addition = '';
            }
            return (index === state.photoTypes.length - 1) ? `and ${photo}` : `${photo}${addition}`;
        }, [])
        .join(' ');
    const finalText = newText.length > 0 ? newText : "nothing";
    setText(`The ultimate repository, currently showing ${finalText}`);
}, [state.photoTypes, setText])
```